### ISSUE TITLE
OOM issue when facing multiple GPU training in stage2

### ISSUE THREAD

**Issue by ranchen293** [haoningwu3639/StoryGen#48]
When I was using the official command-line multi-card training for the second stage, the same four 3090 graphics cards did not run properly and an error OOM was reported. I also tried running the official configuration with only one card, but it still caused OOM. At the same time, I found that when the batchisize drops to 8, running with 4 cards is the same speed as running with 1 card, just as slow.

Could you give me some suggestions and solutions?

**Comment by haoningwu3639** [comment]
Regarding your questions, I can offer the following suggestions:

1. If you encounter OOM, first check whether your computational optimization dependencies, like *xformers* and *accelerate* are working properly.
2. You can offload some frozen parameters, such as **text encoder** and **VAE**, by extracting their corresponding features in advance.
3. OOM sometimes occurs during the VAE's image decoding phase in inference. Consider emptying the training cache before inference.
4. Reduce the batch size and increase the gradient accumulation steps while keeping the total batch size consistent. This trades time overhead for memory overhead.
5. In multi-GPU training, the computation time per GPU is similar to single-GPU, but your total batch size is larger. Thus, you can use fewer gradient accumulation steps to achieve faster training speeds.

**Comment by ranchen293** [comment]
Thank you, project author, for your answers to my previous questions. However, I still have some questions to ask regarding these answers:

1. xformers and accelerate are both working properly.
2. Are there any reference codes to implement cache latents? The official code does not have this step.
3. At present, we haven't considered the inference stage yet, and OOM has already occurred before the training even begins in stage 2.
4. When I execute multi-GPU commands, by checking the progress bar of the terminal, I feel that only one gpu is training, but the video memory of all four gpus is fully utilized. According to your opinion, in this case, is the final batchsize= 4 * the batchsize in config?
5. Our experimental environment also has four 3090s, but it cannot run through the code of multi gpu.

I would be sincerely grateful if you could continue to offer me suggestions!

**Comment by haoningwu3639** [comment]
The batch size specified in the config refers to the per-GPU batch size. Therefore, when using multi-GPU training, the total batch size should be [GPU number] × [config batch size]. I recommend first reducing your batch size (e.g., to 1) and increasing gradient checkpointing steps to ensure your environment can properly execute the training code. Then, you can try to find the optimal trade-off between batch size and gradient checkpointing steps across multiple GPUs.
